### PR TITLE
Add support for Waveshare 7.5inch HD e-paper

### PR DESCRIPTION
**Pull request by nicko029**
## Description:
Add support for Waveshare 7.5inch HD e-paper

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#764

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by jesserockz — APPROVED** [review]

**Comment by jesserockz** [comment]
Just awaiting the docs PR to be fixed to the right branch.

**Comment by grischard** [comment]
@nicko029 thank you for doing the heavy lifting on this. I can't get it to display anything - were you able to get it to work yourself?

**Comment by nicko029** [comment]
@grischard no problem ;) I couldn’t test it, due to the faulty display, so I didn’t know if there is a bug in the code or the display is dead :/

**Comment by grischard** [comment]
Thanks! My display works with the example code from waveshare, but stays completely blank with this in esphome with your PR:

```yaml
esphome:
  name: e_ink_display
  platform: ESP32
  board: esp32dev

spi:
  clk_pin: 13
  mosi_pin: 14

display:
  - platform: waveshare_epaper
    cs_pin: 15
    busy_pin: 25
    reset_pin: 26
    dc_pin: 27
    model: 7.50inhd
    lambda: |-
        ESP_LOGI("display", "Updating display");
        it.line(0, 0, 100, 50);
        it.filled_rectangle(100, 50, 100, 100);
        it.rectangle(200, 150, 30, 42);

logger:
  level: VERBOSE

wifi:
  ap:
    ssid: E-ink Display Fallback Hotspot
    password: !secret 'ap_wifi_password'
  networks:
  - ssid: !secret 'wifi_ssid'
    password: !secret 'wifi_password'
ota:
web_server:
```

File: esphome/components/waveshare_epaper/waveshare_epaper.cpp [hunk 1]
```diff
@@ -888,5 +888,87 @@ void WaveshareEPaper7P5InV2::dump_config() {
   LOG_PIN("  Busy Pin: ", this->busy_pin_);
   LOG_UPDATE_INTERVAL(this);
 }
+
+void WaveshareEPaper7P5InHD::initialize() {
```
Comment (kbx81) [inline esphome/components/waveshare_epaper/waveshare_epaper.cpp:892]: From here down to line 946...can we get some constants defined for all this stuff? I have no idea what any of this accomplishes.

Also, why were the comments removed? They're helpful to understand what's supposed to be going on...

File: esphome/components/waveshare_epaper/waveshare_epaper.cpp [hunk 2]
```diff
@@ -888,5 +888,87 @@ void WaveshareEPaper7P5InV2::dump_config() {
...
+  }
+
+  // COMMAND DISPLAY REFRESH
+  this->command(0x12);
+  delay(100);  // NOLINT
+  this->wait_until_idle_();
+}
+
+int WaveshareEPaper7P5InHD::get_width_internal() { return 880; }
+int WaveshareEPaper7P5InHD::get_height_internal() { return 528; }
```
Comment (kbx81) [inline esphome/components/waveshare_epaper/waveshare_epaper.cpp:964]: Again, use (return) constants please.

**Review by kbx81 — CHANGES_REQUESTED** [review]
Just suggesting readability changes -- nothing major. Add some constants and restore/rewrite comments.

**Comment by grischard** [comment]
Awful, terrible, unreadable specification at https://www.waveshare.com/w/upload/2/27/7inch_HD_e-Paper_Specification.pdf

**Comment by kbx81** [comment]
Here are the register definitions I came up with based on that manual -- hopefully this helps speed along updating of the init code (so it can at least be somewhat more readable).

```c
// 7.5inch HD e-Paper commands
static const uint8_t HDEPAPER75_DRIVER_OUTPUT_CONTROL = 0x01;
static const uint8_t HDEPAPER75_GATE_DRIVING_VOLTAGE = 0x03;
static const uint8_t HDEPAPER75_SOURCE_DRIVING_VOLTAGE = 0x04;
static const uint8_t HDEPAPER75_GATE_SCAN_START_POSITION = 0x0F;
static const uint8_t HDEPAPER75_DEEP_SLEEP_MODE = 0x10;
static const uint8_t HDEPAPER75_DATA_ENTRY_MODE = 0x11;
static const uint8_t HDEPAPER75_SW_RESET = 0x12;
static const uint8_t HDEPAPER75_HV_READY = 0x14;
static const uint8_t HDEPAPER75_VCI_DETECT = 0x15;
static const uint8_t HDEPAPER75_TEMP_SENSOR_SEL = 0x18;
static const uint8_t HDEPAPER75_TEMP_SENSOR_WRITE = 0x1A;
static const uint8_t HDEPAPER75_TEMP_SENSOR_READ = 0x1B;
static const uint8_t HDEPAPER75_TEMP_SENSOR_WRITE_EXT = 0x1C;
static const uint8_t HDEPAPER75_MASTER_ACTIVATION = 0x20;
static const uint8_t HDEPAPER75_DISPLAY_UPDATE_CONTROL_1 = 0x21;
static const uint8_t HDEPAPER75_DISPLAY_UPDATE_CONTROL_2 = 0x22;
static const uint8_t HDEPAPER75_WRITE_RAM_BLACK = 0x24;
static const uint8_t HDEPAPER75_WRITE_RAM_RED = 0x26;
static const uint8_t HDEPAPER75_READ_RAM = 0x27;
static const uint8_t HDEPAPER75_VCOM_SENSE = 0x28;
static const uint8_t HDEPAPER75_VCOM_SENSE_DURATION = 0x29;
static const uint8_t HDEPAPER75_PROGRAM_VCOM_OTP = 0x2A;
static const uint8_t HDEPAPER75_WRITE_REGISTER_VCOM = 0x2B;
static const uint8_t HDEPAPER75_WRITE_VCOM = 0x2C;
static const uint8_t HDEPAPER75_READ_OTP = 0x2D;
static const uint8_t HDEPAPER75_CRC_CALC = 0x34;
static const uint8_t HDEPAPER75_READ_CRC_STATUS = 0x35;
static const uint8_t HDEPAPER75_PROGRAM_OTP = 0x36;
static const uint8_t HDEPAPER75_OTP_DISPLAY_OPTION_REG = 0x37;
static const uint8_t HDEPAPER75_OTP_USER_ID_REG = 0x38;
static const uint8_t HDEPAPER75_OTP_PROGRAM_MODE = 0x39;
static const uint8_t HDEPAPER75_RESERVED1 = 0x3A;
static const uint8_t HDEPAPER75_RESERVED2 = 0x3B;
static const uint8_t HDEPAPER75_BORDER_WAVEFORM = 0X3C;
static const uint8_t HDEPAPER75_READ_RAM_OPTION = 0x41;
static const uint8_t HDEPAPER75_SET_RAM_X_START_END_POS = 0x44;
static const uint8_t HDEPAPER75_SET_RAM_Y_START_END_POS = 0x45;
static const uint8_t HDEPAPER75_AUTO_WRITE_RED_RAM = 0x46;
static const uint8_t HDEPAPER75_AUTO_WRITE_BLACK_RAM = 0x47;
static const uint8_t HDEPAPER75_SET_RAM_X_ADDRESS = 0x4E;
static const uint8_t HDEPAPER75_SET_RAM_Y_ADDRESS = 0x4F;
static const uint8_t HDEPAPER75_NOP = 0x7F;
```

This is more typical of what we have for other displays in the codebase.

File: esphome/components/waveshare_epaper/waveshare_epaper.cpp (hunk 1 above)
Comment (kbx81) [reply esphome/components/waveshare_epaper/waveshare_epaper.cpp:892]: Please see the comment I left below for register defs that should help here.

**Comment by grischard** [comment]
Also maybe useful, from the good people of the GxEPD2 project:

https://github.com/ZinggJM/GxEPD2/blob/master/src/epd3c/GxEPD2_750c_Z90.h

https://github.com/ZinggJM/GxEPD2/blob/master/src/epd3c/GxEPD2_750c_Z90.cpp

**Comment by luigi311** [comment]
Looks like this PR was still missing some commands on the display function so it wasn't displaying anything. I created a PR to his branch that has the missing commands and I was able to get things displaying. I get a timeout error on the log but I think that's due to them changing the readbusy command. If you are using docker you can point your image to luigi311/esphome:waveshare to test my build.

**Comment by grischard** [comment]
@luigi311 I'm working on this too! https://github.com/grischard/esphome/tree/waveshare7.50inhd is where I'm at, and it's working more or less to display things in black. I'd like to be able to display things in both black and red before creating my own PR. Let's work on it together? Are you on the discord chat?

**Comment by jesserockz** [comment]
Added `blocked` due to comments from others that it is not working yet.

**Pull request closed without merging**